overlord/snapstate/autorefresh_gating.go

```diff
@@ -0,0 +1,201 @@
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/release"
+	"github.com/snapcore/snapd/snap"
+)
+
+var gateAutoRefreshHookName = "gate-auto-refresh"
+
+type affectedSnapInfo struct {
+	Restart        bool
+	Base           bool
+	AffectingSnaps map[string]bool
+}
+
+func affectedByRefresh(st *state.State, updates []*snap.Info) (map[string]*affectedSnapInfo, error) {
+	all, err := All(st)
+	if err != nil {
+		return nil, err
+	}
+
+	var bootBase string
+	if !release.OnClassic {
+		deviceCtx, err := DeviceCtx(st, nil, nil)
+		if err != nil {
+			return nil, fmt.Errorf("cannot get device context: %v", err)
+		}
+		bootBaseInfo, err := BootBaseInfo(st, deviceCtx)
+		if err != nil {
+			return nil, fmt.Errorf("cannot get boot base info: %v", err)
+		}
+		bootBase = bootBaseInfo.InstanceName()
+	}
+
+	byBase := make(map[string][]string)
+	for name, snapSt := range all {
+		if !snapSt.Active {
+			delete(all, name)
+			continue
+		}
+		inf, err := snapSt.CurrentInfo()
+		if err != nil {
+			return nil, err
+		}
+		// optimize: do not consider snaps that don't have gate-auto-refresh hook.
+		if inf.Hooks[gateAutoRefreshHookName] == nil {
+			delete(all, name)
+			continue
+		}
+
+		base := inf.Base
+		if base == "none" {
+			continue
+		}
+		if inf.Base == "" {
+			base = "core"
+		}
+		byBase[base] = append(byBase[base], inf.InstanceName())
+	}
+
+	affected := make(map[string]*affectedSnapInfo)
+
+	addAffected := func(snapName, affectedBy string, restart bool, base bool) {
+		affectedInfo := affected[snapName]
+		if affectedInfo == nil {
+			affectedInfo = &affectedSnapInfo{
+				AffectingSnaps: map[string]bool{},
+			}
+			affected[snapName] = affectedInfo
+		}
+		if restart {
+			affectedInfo.Restart = restart
+		}
+		if base {
+			affectedInfo.Base = base
+		}
+		affectedInfo.AffectingSnaps[affectedBy] = true
+	}
+
+	for _, up := range updates {
+		// on core system, affected by update of boot base
+		if bootBase != "" && up.InstanceName() == bootBase {
+			for _, snapSt := range all {
+				addAffected(snapSt.InstanceName(), up.InstanceName(), true, true)
+			}
+		}
+
+		// snaps that can trigger reboot
+		// XXX: gadget refresh doesn't always require reboot, refine this
+		if up.Type() == snap.TypeKernel || up.Type() == snap.TypeGadget {
+			for _, snapSt := range all {
+				addAffected(snapSt.InstanceName(), up.InstanceName(), true, false)
+			}
+			continue
+		}
+		if up.Type() == snap.TypeBase || up.SnapName() == "core" {
+			// affected by refresh of this base snap
+			for _, snapName := range byBase[up.InstanceName()] {
+				addAffected(snapName, up.InstanceName(), false, true)
+			}
+			continue
+		}
+
+		repo := ifacerepo.Get(st)
+
+		// consider slots provided by refreshed snap, but exclude core and snapd
+		// since they provide system-level slots that are generally not disrupted
+		// by snap updates.
+		if up.SnapType != snap.TypeSnapd && up.SnapName() != "core" {
```

Comment:
we don't need the 2nd check for core because of the continue at 126? but maybe is too subtle?

Reply:
Indeed, it's excessive. I added a comment and removed it.